src/compiler/scala/tools/nsc/classpath/ZipAndJarFileLookupFactory.scala

```diff
@@ -215,3 +300,17 @@ final class FileBasedCache[T] {
     cache.clear()
   }
 }
+
+object FileBasedCache {
+  // The tension here is that too long a delay could lead to an error (on Windows) with an inability
+  // to overwrite the JAR. To short a delay and the entry could be evicted before a subsequent
+  // sub-project compilation is able to get a cache hit. A more comprehensive solution would be to
+  // involve build tools in the policy: they could close entries with refcount of zero when that
+  // entry's JAR is about to be overwritten.
+  private val deferCloseMs = Integer.getInteger("scalac.filebasedcache.defer.close.ms", 1000)
```

Comment:
From my side, I'd like a mechanism to enable this class loader caching but be in charged of the invalidation. I feel it would be better to expose an API here instead of relying on concrete top-level caches that cannot be modified easily from the outside.
    
Can we add an API that we can override in global that allows us to handle the lifetime of these caches?
  
With this API, I'd be able to override it in Zinc and use it in Bloop to correctly invalidate these caches and free up the file pointers when I know a user operation will write to an open cached file (e.g. when the project of a compiler plugin is recompiled, when a dependent project that uses straight-to-jar compilation is used, etc etc). This will enable 100% correct usage of class loader caching with zero user intervention, which is what I'm aiming for.

Reply:
My idea was that you could override `findMacroClassLoader` and `findPluginClassloader` and DIY caching / invalidation.

Reply:
Now that this is already present in analyser I'm confident we'll be able to implement our own invalidation. Thanks!